pymagicc/io.py

```diff
@@ -1344,6 +1342,164 @@ def find_region(x):
     return region_dattype_row
 
 
+class _CompactOutReader(_Reader):
+    def read(self):
+        compact_table = self._read_compact_table()
+
+        return self._convert_compact_table_to_df_metadata_column_headers(compact_table)
+
+    def _read_compact_table(self):
+        with open(self.filepath, "r") as fh:
+            headers = self._read_header(fh)
+            # can change to reading limited number of lines in future
+            lines_as_dicts = [l for l in self._read_lines(fh, headers)]
+
+        return pd.DataFrame(lines_as_dicts)
+
+    def _read_header(self, fh):
+        l = fh.readline().strip(",\n")
+        return [item.strip('"') for item in l.split(",")]
+
+    def _read_lines(self, fh, headers):
+        for line in fh:
+            toks = line.strip(",\n").split(",")
+
+            assert len(toks) == len(headers), "# headers does not match # lines"
+
+            yield {h: float(v) for h, v in zip(headers, toks)}
+
+    def _convert_compact_table_to_df_metadata_column_headers(self, compact_table):
+        ts_cols = [c for c in compact_table if "__" in c]
+        para_cols = [c for c in compact_table if "__" not in c]
+
+        ts = compact_table[ts_cols]
+        ts = ts.T
+
+        def sort_ts_ids(inid):
+            variable, region, year = inid.split("__")
+            variable = variable.replace("DAT_", "")
+
+            return {"variable": variable, "region": region, "year": year}
+
+        ts["variable"] = ts.index.map(
+            lambda x: convert_magicc7_to_openscm_variables(
+                x.split("__")[0].replace("DAT_", "")
+            )
+        )
+        ts["region"] = ts.index.map(
+            lambda x: convert_magicc_to_openscm_regions(x.split("__")[1])
+        )
+        ts["year"] = ts.index.map(lambda x: x.split("__")[2])
+        if not (ts["year"].apply(len) == 4).all():  # pragma: no cover # safety valve
```

Comment:
This is checking the length of an int cast to a string right? It might be clearer to cast the years to ints in 1 line and then check for uniqueness

```
ts["year"] = ts.index.map(lambda x: x.split("__")[2]).astype(int)
if not ts["year"].is_unique:
    explode
```

Reply:
yep good suggestion

Reply:
I lie, at this point the years won't be unique cause you'll have e.g. `DAT_CO2_CONC_1765` and `DAT_SURFACE_TEMP_1765`.